wptrunner/executors/pytestrunner/fixtures.py

```diff
@@ -45,14 +46,66 @@ def setup(request, session):
         def test_something(setup, session):
             assert session.url == "https://example.org"
 
-    The session is closed when the test module goes out of scope by an
-    implicit call to `session.end`.
+    When the test function goes out of scope, any remaining user prompts
+    and opened windows are closed, and the current browsing context is
+    switched back to the top-level browsing context.
     """
 
     def __init__(self, client):
         self.client = client
 
-    @pytest.fixture(scope="module")
+    @pytest.fixture(scope="function")
     def session(self, request):
-        request.addfinalizer(self.client.end)
+        # finalisers are popped off a stack,
+        # making their ordering reverse
+        request.addfinalizer(self.switch_to_top_level_browsing_context)
+        request.addfinalizer(self.restore_windows)
+        request.addfinalizer(self.dismiss_user_prompts)
+
         return self.client
+
+    def dismiss_user_prompts(self):
+        """Dismisses any open user prompts in windows."""
+        current_window = self.client.window_handle
+
+        for window in self.windows():
+            self.client.window_handle = window
+            try:
+                self.client.alert.dismiss()
+            except webdriver.NoSuchAlertException:
+                pass
+
+        self.client.window_handle = current_window
+
+    def restore_windows(self):
+        """Closes superfluous windows opened by the test without ending
+        the session implicitly by closing the last window.
+        """
+        for window in self.windows():
```

Comment:
Since the dismiss user prompts seems to take care of tracking the current window handle, shouldn't we try to exclude it from being closed here? Since the order returned by getting all windows isn't guaranteed, but we shouldn't assume the current window is actually in the set